cpp/src/arrow/dataset/partition.h

```diff
@@ -108,10 +135,25 @@ class ARROW_DS_EXPORT PartitionScheme {
 /// the form $key=$value in directory names
 class ARROW_DS_EXPORT HivePartitionScheme : public PartitionScheme {
  public:
-  /// \brief Return true if path
-  bool PathMatchesScheme(const std::string& path) const override;
+  explicit HivePartitionScheme(std::shared_ptr<Schema> schema)
+      : schema_(std::move(schema)) {}
+
+  std::string name() const override { return "hive_partition_scheme"; }
 
-  virtual Status ParseKey(const std::string& path, PartitionKeyData* out) const = 0;
+  Status Parse(const std::string& path, std::string* unconsumed,
+               std::shared_ptr<Expression>* out) const override;
+
+  std::vector<UnconvertedKey> GetUnconvertedKeys(const std::string& path,
+                                                 std::string* unconsumed) const;
+
+ protected:
+  /// XXX do we have a schema when constructing partition schemes?
```

Comment:
The user is usually aware of the types and names of the keys. It should be passed at construction time. If none is passed, we should default to string types.

Reply:
Defaulting to string seems like a foot gun to me. Let's require the schema for now. I'll remove this comment